src/main/scala/vectorpipe/vectortile/Pipeline.scala

```diff
@@ -0,0 +1,126 @@
+ * reduce]] and [[simplify]] will be called prior to processing the initial zoom
+ * level.
+ */
+trait Pipeline {
+  /**
+   * The root URI for output.
+   */
+  val baseOutputURI: java.net.URI
+
+  /**
+   * Name of the column containing the JTS geometry.
+   */
+  val geometryColumn: String
+
+  /**
+   * Reduce the input data between zoom levels.
+   *
+   * Not all data is useful in all levels of the vector tile pyramid.  When a
+   * geometry will no longer be used in any subsequent generated level, it may be
+   * desirable to drop the relevant rows from the DataFrame to limit the amount
+   * of working data.  Moreover, it may be desirable to collect multiple entities
+   * into a single, aggregated entity.  These sorts of operations are made
+   * possible by the [[reduce]] function.
+   *
+   * This function will be called once before tile generation begins at the
+   * initial zoom level.  All participating geometries will already have been
+   * keyed to the given layout, with the list of relevant keys available in the
+   * "keys" field.
+   *
+   * @param   input           A DataFrame minimally containing a "geom" field of
+   *                          JTS [[Geometry]] and a field of Array[SpatialKey]
+   *                          with the name given by the keyColumn param, below
+   * @param   layoutLevel     The layout level for the target zoom
+   * @param   keyColumn       The name of the column containing
+   *                          Array[SpatialKey] giving the list of keys that the
+   *                          geometry interacts with
+   */
+  def reduce(input: DataFrame, layoutLevel: LayoutLevel, keyColumn: String): DataFrame = input
+
+  /*
+   * Lower complexity of geometry while moving to less resolute zoom levels.
+   *
+   * While moving from finer to coarser levels of the pyramid, it may not be
+   * necessary to maintain the full level of detail of available geometries.
+   * This function is used to reduce the complexity of geometries.  The
+   * [[LayoutDefinition]] will be for the target zoom level.
+   *
+   * By default, we return the input geometry without simplification, but there
+   * is a simplifier using JTS's topology-preserving simplifier available in
+   * [[vectorpipe.vectortile.Simplify]].
+   */
+  def simplify(g: jts.Geometry, layout: LayoutDefinition): jts.Geometry = g
+
+  /**
+   * Select geometries for display at a given zoom level.
+   *
+   * VectorPipe allows geometries to be carried up the pyramid for display at a
+   * later time.  This function is used to choose the elements that will be
+   * displayed at the target zoom level.  This is useful for implementing a
+   * display schema, or for example, selection based on prominence (relying on
+   * the array of spatial keys added to the data frame in the column with the
+   * longest name fitting the regular expression '[_]+keys').
+   */
+  def select(input: DataFrame, targetZoom: Int): DataFrame = input
```

Comment:
My gut says that the `DataFrame` returned by this should include a layer name (`_layer`?) with each row, in order to curate features into different layers (and potentially to filter / modify their attributes accordingly).

If I have a heterogenous collection of roads, buildings and POIs, I'd like to be able to reclassify the highways into fewer classifications, keep a limited set of tags according to type, and render each category into a different layer.

I'm thinking something like this:

```scala
  // tags to keep by type
  val POITags = Set("amenity", "name")
  val BuildingTags = Set("addr:number", "amenity", "name")
  val RoadTags = Set("highway", "name")

  val rewriteTags: UserDefinedFunction = udf { (category: String, tags: Map[String, String]) =>
    category match {
      case "road" =>
        tags.filterKeys(k => RoadTags.contains(k)) ++ Map("type" -> reclassifyRoad(tags("highway")))

      case "building" =>
        tags.filterKeys(k => BuildingTags.contains(k))

      case "poi" =>
        tags.filterKeys(k => POITags.contains(k))
    }
  }

  def categorize(tags: Column): Column =
    when(isRoad(tags), "road")
      .when(isBuilding(tags), "building")
      .when(isPOI(tags), "poi") as 'category

  // group OSM classifications into a simplified hierarchy
  def reclassifyRoad(classification: String): String = ???

  changes
    // group features into layers by category
    .withColumn("_layer", categorize('tags))
    // drop uncategorized features
    .where('_layer.isNotNull)
    // rewrite tags
    .withColumn("tags", rewriteTags('_layer, 'tags))
```

Reply:
This is probably the best idea going.  I'll give it a shot.  Probably best to ignore the (mostly naïve) comment about single geometries in multiple layers.  I suppose you could easily generate multiple layer tags and explode them in select and solve that problem.

Reply:
Individual features in multiple layers seems uncommon, but I can imagine it happening (though I am having a hard time coming up with a contrived use-case). Exploding would definitely work in a custom `select` implementation though.